clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3047,7 +3062,7 @@ class SYCLKernelNameTypeVisitor
           IsInvalid = true;
           return;
         }
-        if (Tag->isCompleteDefinition()) {
+        if (isa<FunctionDecl>(DeclCtx) && Tag->isCompleteDefinition()) {
```

Comment:
You're correct in a class inside a function scope is not globally visible.  So when the 'visitor' finds a function-type/decl, I would assume it would reject it.

Globally visible names appear only in namespaces, record types, or enums I believe (though I don't believe you can create a record-type inside an enum in any way, unless there is an incredible trick I'm missing).

Reply:
>>I believe I don't know exactly what it means, but I asked Sri to clarify it before preparing a patch.
I looked at "5.2 Naming of kernels" in the SYCL2020-provisional-spec and based on my understanding of rule#2,3,4 made this change.